controllers/internalrequest/adapter.go

```diff
@@ -18,34 +18,132 @@ package internalrequest
 		client:          client,
-		context:         context,
+		ctx:             ctx,
 		internalRequest: internalRequest,
 		internalClient:  internalClient,
+		loader:          loader,
 		logger:          logger,
 	}
 }
 
-func (a *Adapter) EnsureReconcileOperationIsLogged() (reconciler.OperationResult, error) {
-	a.logger.Info("InternalRequest successfully watched")
+// EnsurePipelineRunIsCreated is an operation that will ensure that the InternalRequest is handled by creating a new
+// PipelineRun for the Pipeline referenced in the Request field.
+func (a *Adapter) EnsurePipelineRunIsCreated() (reconciler.OperationResult, error) {
+	pipelineRun, err := a.loader.GetInternalRequestPipelineRun(a.ctx, a.internalClient, a.internalRequest)
+	if err != nil && !errors.IsNotFound(err) {
+		return reconciler.RequeueWithError(err)
+	}
+
+	if pipelineRun == nil || !a.internalRequest.HasStarted() {
+		if pipelineRun == nil {
+			pipelineRun = tekton.NewPipelineRun(a.internalRequest)
+
+			err = libhandler.SetOwnerAnnotations(a.internalRequest, pipelineRun)
+			if err != nil {
+				return reconciler.RequeueWithError(err)
+			}
+
+			err = a.internalClient.Create(a.ctx, pipelineRun)
+			if err != nil {
+				return reconciler.RequeueWithError(err)
+			}
+
+			a.logger.Info("Created PipelineRun to handle request",
+				"PipelineRun.Name", pipelineRun.Name, "PipelineRun.Namespace", pipelineRun.Namespace)
+		}
+
+		return reconciler.RequeueOnErrorOrContinue(a.registerInternalRequestStatus(pipelineRun))
+	}
 
 	return reconciler.ContinueProcessing()
 }
+
+// EnsureStatusIsTracked is an operation that will ensure that the release PipelineRun status is tracked
+// in the InternalRequest being processed.
+func (a *Adapter) EnsureStatusIsTracked() (reconciler.OperationResult, error) {
+	pipelineRun, err := a.loader.GetInternalRequestPipelineRun(a.ctx, a.internalClient, a.internalRequest)
+	if err != nil && !errors.IsNotFound(err) {
+		return reconciler.RequeueWithError(err)
+	}
+
+	if pipelineRun != nil {
+		return reconciler.RequeueOnErrorOrContinue(a.registerInternalRequestPipelineRunStatus(pipelineRun))
+	}
+
+	return reconciler.ContinueProcessing()
+}
+
+// registerInternalRequestStatus sets the InternalRequest to Running.
+func (a *Adapter) registerInternalRequestStatus(pipelineRun *v1beta1.PipelineRun) error {
+	if pipelineRun == nil {
+		return nil
+	}
+
+	patch := client.MergeFrom(a.internalRequest.DeepCopy())
+
+	a.internalRequest.MarkRunning()
+
+	return a.client.Status().Patch(a.ctx, a.internalRequest, patch)
+}
+
+// registerInternalRequestPipelineRunStatus keeps track of the PipelineRun status in the InternalRequest being processed.
+func (a *Adapter) registerInternalRequestPipelineRunStatus(pipelineRun *v1beta1.PipelineRun) error {
+	if pipelineRun == nil || !pipelineRun.IsDone() {
+		return nil
+	}
+
+	patch := client.MergeFrom(a.internalRequest.DeepCopy())
+
+	condition := pipelineRun.Status.GetCondition(apis.ConditionSucceeded)
+	if condition.IsTrue() {
```

Comment:
In python, this would be if - elif - else. I don't know what the conventions are in Go for this, but would it make sense to format it like this?

```
if condition.IsTrue() {
    ...
} else if strings.Contains(condition.Message, "not found") {
    ...
} else {
    ...
}
```

Reply:
That works as well. I did it as it is to have a clear distinction between true and false conditions, but I'm ok with this code as well.